File: Python/twoSum.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: twoSum.py
@Function: 两数之和
@Link: https://leetcode-cn.com/problems/two-sum/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-04
"""

# --------------------------------------
class Solution1:
    """
    暴力穷举法，使用两个 for 循环即可实现
    遍历每个元素 x，并查找是否存在一个值与 target - x 相等的目标元素
    时间复杂度：big O(n**2)
    空间复杂度：big O(1)

    仔细观察，时间复杂度和空间复杂度权衡
    时间复杂度的提升是以空间复杂度为代价的
    """

    def twoSum(self, nums, target):
        index = []
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i,j]
    

# ---------------------------------
class Solution2:
    """
    元素和下标，键值对方式储存,字典数据结构
    遍历键值对的键 key，并查找是否存在一个值与 target - key 相等的键值对应的value即为所求下标
    时间复杂度：big O(n)
    空间复杂度：big O(1)

    仔细观察，时间复杂度和空间复杂度权衡
    时间复杂度的提升是以空间复杂度为代价的
    """

    def twoSum(self, nums, target):
        map_dict = dict()
        ret = list()
        for item in range(len(nums)):
            map_dict[nums[item]] = item
        
        for i in range(len(nums)):
            temp = target - nums[i]
            if temp in map_dict.keys() and map_dict[temp] != i:
                ret.append(i)
                ret.append(map_dict[temp])
                break;

        return ret


# ------------------------------
class Solution3:
    """
    enumerate 遍历 list，
    将一个可遍历的数据对象(如列表、元组或字符串)
    组合为一个索引序列，同时列出数据和数据下标，一般用在 for 循环当中 
    
    组合数据和数据下标思想！！！
    算法核心思想与 solution2 一致，只不过使用Python 内置函数 enumerate 实现
    """
    def twoSum(self, nums, target):
        dict = {}
        # enumerate 内置函数，常用 for 循环遍历一个容器/迭代器
        # 参数 sequence -- 一个序列、迭代器或其他支持迭代对象
        # 参数 start -- 下标起始位置
        # 返回值 返回 enumerate(枚举) 对象
        for i,n in enumerate(nums,0):
            if target - n in dict:
                return [dict[target-n],i]
            dict[n] = i
             
# -----------------------------------------------------
# The key to the problem is that there is ALWAYS only 1 pair of numbers that satisfy the condition of adding together to be the target value.
# We can assume that for all the numbers in the list (x1, x2, ... xn) that there exists a pair such that xa + xb = target
# To solve this with a single pass of the list we can change the equation above to xa = target - xb and since we know the target as long as we maintain a record of all previous values in the list we can compare the current value (xa) to it's ONLY pair, if it exists, in record of all previous values (xb)

# To keep a record of the previous values and their indices I have used a dictionary. Commonly known as a map in other languages. This allows me to record each previous number in the dictionary alongside the indice as a key value pair (target-number, indice).
# -----------------------------------------------------
# https://leetcode.com/problems/can-place-flowers/
# https://leetcode-cn.com/problems/can-place-flowers/
# 可以查看高手的解答，学习，(中文的-cn/), 国外的高手解答，选择【 Most Votes 】
# -----------------------------------------------------


# --------------------------------
if __name__ == "__main__":
    nums = [2, 7, 11, 15]
    target = 9
    # solution = Solution1()
    # solution = Solution2()
    solution = Solution3()
    print(solution.twoSum(nums, target))

File: C++/twoSum.cpp
/**
 * @File    : twoSum.cpp
 * @Brief   : 两数之和
 * @Link    : https://leetcode-cn.com/problems/two-sum/
 * @Author  : Wei Li
 * @Date    : 2021-07-04
*/

// -----------------------------------------------------------------------
// 1、指针与常量
// int x;
// int * p1 = &x; // 指针可以被修改， 值也可以被修改
// const int * p2 = &x; // 指针可以被修改， 值不可以被修改（const int）
// int * const p3 = &x; // 指针不可以被修改（* const）， 值可以被修改
// const int * const p4 = &x; // 指针不可以被修改， 值也不可以被修改

// 2、指针函数与函数指针
// addition是指针函数， 一个返回类型是指针的函数
// int* addition(int a, int b) 
// {
//     int* sum = new int(a + b);
//     return sum;
// }

// int subtraction(int a, int b) 
// {
//     return a - b;
// }

// int operation(int x, int y, int (*func)(int, int)) 
// {
//     return (*func)(x,y);
// }

// // minus是函数指针， 指向函数的指针
// int (*minus)(int, int) = subtraction;
// int* m = addition(1, 2);
// int n = operation(3, *m, minus);
// -----------------------------------------------------------------------


// [include]
#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>


// -----------------------------
class Solution1
{
public:
    static std::vector<int> twoSum(std::vector<int> &nums, int target)
    {
        /*
        ** 暴力穷举法，使用两个 for 循环即可实现
        ** 遍历每个元素 x，并查找是否存在一个值与 target - x 相等的目标元素
        ** 时间复杂度：big O(n**2)
        ** 空间复杂度：big O(1)
        **
        ** 仔细观察，时间复杂度和空间复杂度权衡
        ** 时间复杂度的提升是以空间复杂度为代价的
        */

        for (size_t i = 0; i < nums.size(); ++i)
        {
            for (size_t j = i + 1; j < nums.size(); ++j)
            {
                if (nums[j] == target - nums[i])
                {
                    std::vector<int> iVec;
                    iVec.push_back(i);
                    iVec.push_back(j);
                    return iVec;
                }
                else
                {
                    continue;
                }
            }
        }
        throw("No two sum solution");
    }
};


// -----------------------------------------
class Solution2
{
public:
    static std::vector<int> twoSum(std::vector<int> &nums, int target)
    {
        /*
        ** 元素和下标，键值对方式储存
        ** 遍历键值对的键 key，并查找是否存在一个值与 target - key 相等的键值对应的value即为所求下标
        ** 时间复杂度：big O(n)
        ** 空间复杂度：big O(1)
        **
        ** 仔细观察，时间复杂度和空间复杂度权衡
        ** 时间复杂度的提升是以空间复杂度为代价的
        */

        std::map<int, int> iMap;
        std::vector<int> ret;
        for (int i = 0; i < nums.size(); ++i)
        {
            // 数组的值作为 map的 key；下标作为 map的 value
            iMap[nums[i]] = i;
        }
        for (int i = 0; i < nums.size(); ++i)
        {
            int temp = target - nums[i];
            // 检索到数组中的值，同时满足同一个元素不能使用两遍
            if (iMap.count(temp) && iMap[temp] != i)
            {
                ret.push_back(i);
                ret.push_back(iMap[temp]);
                break;
            }
        }
        return ret;
    }
};


// -----------------------------------
class Solution3
{
public:
    static std::vector<int> twoSum(std::vector<int> &nums, int target)
    {
        /*
        * 算法核心思想，组合数据和数据的下标为键值对
        * 哈希映射 hash_map
        * 还是对所有数据进行扫描，然后进行对比处理
        * 利用 C11 standard with unordered_map class 进行处理，本身语言的特性
        */

        std::vector<int> result;
        // unordered_map for C11 standard
        std::unordered_map<int, int> hash_map;
        for (int i = 0; i < nums.size(); ++i)
        {
            // 进栈 push
            result.push_back(i);
            //hash_map[nums[i]] = i;
            if (hash_map.find(target - nums[i]) != hash_map.end())
            {
                result.push_back(hash_map[target - nums[i]]);
                hash_map[nums[i]] = i;
                return result;
            }
            hash_map[nums[i]] = i;
            // 出栈 pop
            result.pop_back();
        }
        return result;
    }
};


// -----------------------------------
int main(int argc, char **argv)
{
    // [test]
    std::vector<int> nums{-1, -2, -3, -4, -5};
    int target = -8;

    // vector<int> ivec = Solution1::twoSum(nums, target);
    // vector<int> ivec = Solution2::twoSum(nums, target);
    std::vector<int> ivec = Solution3::twoSum(nums, target);

    std::cout << "the index of two integer : ";
    for (auto iter = ivec.begin(); iter != ivec.end(); ++iter)
    {
        std::cout << *iter << "\t";
    }
    std::cout << std::endl;
    // [test]

    return 0;
}
